random/imx6ull-rngb.c

```diff
@@ -0,0 +1,306 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <posix/utils.h>
+#include <sys/debug.h>
+#include <sys/platform.h>
+#include <sys/interrupt.h>
+#include <sys/threads.h>
+#include <sys/mman.h>
+#include <sys/msg.h>
+
+
+#define RNG_VER 0
+#define RNG_CMD 1
+#define RNG_CR  2
+#define RNG_SR  3
+#define RNG_ESR 4
+#define RNG_OUT 5
+
+
+#define RNGB_START_ADDRESS 0x02284000
+#define RNGB_IRQ           (6 + 32)
+
+
+#define ERROR_OCCURRED 3U
+#define MMAP_SIZE      4096
+
+
+#define FIELD(address, top, length) (((*(address)) << (31 - (top))) >> (32 - (length)))
+
+
+/* intr_st indicates status of interrupts */
+static struct {
+	volatile uint32_t *base;
+	volatile uint32_t intr_st;
+	uint32_t port;
+
+	handle_t lock;
+	handle_t cond;
+} common_rngb;
+
+
+static int rngb_intr(unsigned int n, void *arg)
+{
+	uint32_t filed = *(common_rngb.base + RNG_SR);
+
+	/* Self tests are done */
+	if ((filed & 0x10) != 0) {
+		common_rngb.intr_st = 1U;
+	}
+
+	/* Seeding is done*/
+	if ((filed & 0x20) != 0) {
+		common_rngb.intr_st = 2U;
+	}
+
+	/* An error occurred*/
+	if ((filed & (1 << 16)) != 0) {
+		common_rngb.intr_st = ERROR_OCCURRED;
+	}
+
+	// Clear interrupt
+	*(common_rngb.base + RNG_CMD) = 0x20;
+	return 1;
+}
+
+
+static void systemCleanup(int level)
+{
+	if (level > 0) {
+		munmap((void *)common_rngb.base, MMAP_SIZE);
+	}
+
+	if (level > 1) {
+		portDestroy(common_rngb.port);
+	}
+
+	if (level > 2) {
+		resourceDestroy(common_rngb.lock);
+	}
+
+	if (level > 3) {
+		resourceDestroy(common_rngb.cond);
+	}
+}
+
+
+static int systeminit(void)
+{
+	const char devpath[] = "random";
+	int err;
+	oid_t dev;
+
+	common_rngb.base = mmap(NULL, MMAP_SIZE, PROT_READ | PROT_WRITE, MAP_DEVICE | MAP_UNCACHED, OID_PHYSMEM, RNGB_START_ADDRESS);
+	if (common_rngb.base == MAP_FAILED) {
+		printf("rngb: Could not map addr\n");
+		return -1;
+	}
+
+	err = portCreate(&common_rngb.port);
+	if (err != EOK) {
+		printf("rngb: Could not create port\n");
+
+		systemCleanup(1);
+		return -1;
+	}
+
+	dev.port = common_rngb.port;
+	dev.id = 0;
+
+	err = create_dev(&dev, devpath);
+	if (err != EOK) {
+		printf("rngb: Could not create port file %s (err %d)\n", devpath, err);
+
+		systemCleanup(2);
+		return -1;
+	}
+
+	err = mutexCreate(&common_rngb.lock);
+	if (err != EOK) {
+		printf("rngb: Could not create mutex for rngb\n");
+
+		systemCleanup(2);
+		return -1;
+	}
+
+	err = condCreate(&common_rngb.cond);
+	if (err != EOK) {
+		printf("rngb: Could not create cond for rngb\n");
+
+		systemCleanup(3);
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int hardwareinit(void)
+{
+	int err;
+
+	/* Reset and wait until rngb is sleeping */
+	*(common_rngb.base + RNG_CMD) = 0x40;
+	do {
+		err = *(common_rngb.base + RNG_SR) & 0x4;
+	} while (err == 0);
+
+	interrupt(RNGB_IRQ, rngb_intr, NULL, common_rngb.cond, NULL);
+
+	/* Run self test */
+	*(common_rngb.base + RNG_CMD) = 0x1;
+	while (common_rngb.intr_st == 0) {
+		condWait(common_rngb.cond, common_rngb.lock, 0);
```

Comment:
`condWait` should nb called under locked mutex (you can wrap whole loop). I suggest using 1-3 secs timeout in `condWait` just in case here.